txscript/sigcache.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2015 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package txscript
+
+import (
+	"bytes"
+	"crypto/rand"
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// sigInfo represents an entry in the SigCache. Entries in the sigcache are a
+// 3-tuple: (sigCache, sig, pubKey).
+type sigInfo struct {
+	// TODO(roasbeef): Make this a string so it's hash func blocksize
+	// generic?
+	sigHash wire.ShaHash
+	sig     btcec.Signature
+	pubKey  btcec.PublicKey
+}
+
+// SigCache implements an ECDSA signature verification cache with a randomized
+// entry eviction policy. The benefits of SigCache are two fold. Firstly, usage
+// of SigCache mitigates a DoS attack wherein an attack causes a victim's client
+// to hang due to worst-case behavior triggered while processing attacker
+// crafted invalid transactions. A detailed description of the mitigated DoS
+// attack can be found here: https://bitslog.wordpress.com/2013/01/23/fixed-bitcoin-vulnerability-explanation-why-the-signature-cache-is-a-dos-protection/.
+// Secondly, usage of the SigCache introduces a signature verification
+// optimization which speeds up the validation of transactions within a block,
+// if they've already once been seen and verified within the mempool.
+type SigCache struct {
+	sync.RWMutex
+	validSigs  map[sigInfo]struct{}
+	maxEntries int
+}
+
+// NewSigCache creates and initializes a new instance of SigCache. Its sole
+// parameter 'maxEntries' represents the maximum number of entries allowed to
+// exist in the SigCache and any particular moment. If the addition of a new
+// 'sigInfo' entry puts the number of entries in the SigCache over
+// 'maxEntries', then an existing entry is randomly chosen to be evicted.
+func NewSigCache(maxEntries int) *SigCache {
+	return &SigCache{validSigs: make(map[sigInfo]struct{}), maxEntries: maxEntries}
+}
+
+// Exists returns true if an existing entry of 'sig' over 'sigHash' for public
+// key 'pubKey' is found within the SigCache. Otherwise, false is returned.
+//
+// NOTE: This function is safe for concurrent access. Readers won't be blocked
+// unless there exists a writer, adding an entry to the SigCache.
+func (s *SigCache) Exists(sigHash []byte, sig *btcec.Signature, pubKey *btcec.PublicKey) bool {
+	s.RLock()
```

Comment:
Probably not a big deal, but this lock is being held longer than it needs to be since it is only protecting access to the map and defer has non-trivial overhead.  I'd suggest:

```Go
s.RLock()
_, ok := s.validSigs[info]
s.RUnlock()
```